File: Facebook.NET/GraphAPI.cs
using System;
using System.IO;
using System.Linq;
using System.Net;
using System.Runtime.Serialization.Json;

namespace Facebook
{
    public static partial class GraphAPI
    {
        private const string oAuthCodeUri = "https://www.facebook.com/dialog/oauth";
        private const string accessTokenUri = "https://graph.facebook.com/oauth/access_token";
        private const string meUri = "https://graph.facebook.com/me";
        private const string graphUri = "https://graph.facebook.com/";
        
        private static readonly Guid uniqueID = Guid.NewGuid();

        public static string GetAccessTokenUri(string appID, string appSecret, string authCode, string redirectUri)
        {
            return string.Format(
                string.Format("{0}?client_id={0}&redirect_uri={1}&client_secret={2}&code={3}", accessTokenUri),
                appID, redirectUri, appSecret, authCode);
        }

        public static string GetOAuthCodeUri(string appID, string redirectUri)
        {
            string uri = string.Format(
                string.Format("{0}?client_id={0}&redirect_uri={1}&state={2}", oAuthCodeUri),
                appID, redirectUri, uniqueID);
            return uri;
        }
        internal static string GetParameterizedString(string name, params string[] values)
        {
            bool isNullOrEmpty = values == null || values.Count() == 0;

            return isNullOrEmpty == false ?
                string.Format("{0}={1}", name, values.Aggregate(
                    (previous, current) => string.Format("{0},{1}", previous, current)
                )) : null;
        }

        internal static string GetFormattedUri(string uriBase, params string[] parameters)
        {
            UriBuilder builder = new UriBuilder(uriBase);
            if (parameters != null)
            {
                builder.Query = string.Format("{0}", parameters.Aggregate(
                    (previous, current) => 
                        string.IsNullOrEmpty(previous) ? current :
                        string.IsNullOrEmpty(current) ? previous :
                        string.Format("{0}&{1}", previous, current)
                        ));
            }
            return builder.ToString();
        }

        internal static T DeserializeResponse<T>(string uri, string method = "GET")
        {
            T result;
            WebRequest request = HttpWebRequest.Create(uri);
            request.Method = method;
            using (WebResponse response = request.GetResponse())
            {
                using (Stream stream = response.GetResponseStream())
                {
                    DataContractJsonSerializer serializer = new DataContractJsonSerializer(typeof(T));
                    result = (T)serializer.ReadObject(stream);
                    stream.Close();
                }
                response.Close();
            }

            return result;
        }
    }
}

File: Facebook.NET - VB/GraphAPI.vb
Imports System
Imports System.IO
Imports System.Linq
Imports System.Net
Imports System.Runtime.Serialization.Json

Namespace Facebook
	Public NotInheritable Partial Class GraphApi
        Private Const OAuthCodeUri As String = "https://www.facebook.com/dialog/oauth"
		Private Const AccessTokenUri As String = "https://graph.facebook.com/oauth/access_token"
		Private Const MeUri As String = "https://graph.facebook.com/me"
		Private Const GraphUri As String = "https://graph.facebook.com/"

		Private Shared ReadOnly uniqueID As Guid = Guid.NewGuid()

		Public Shared Function GetAccessTokenUri(appID As String, appSecret As String, authCode As String, redirectUri As String) As String
            Return String.Format("{0}?client_id={0}&redirect_uri={1}&client_secret={2}&code={3}",
                          AccessTokenUri,
                          appID,
                          redirectUri,
                          appSecret,
                          authCode
            )
		End Function

		Public Shared Function GetOAuthCodeUri(appID As String, redirectUri As String) As String
            Return String.Format("{0}?client_id={0}&redirect_uri={1}&state={2}",
                                 OAuthCodeUri,
                                 appID,
                                 redirectUri,
                                 uniqueID
            )
		End Function
		Friend Shared Function GetParameterizedString(name As String, ParamArray values As String()) As String
			Dim isNullOrEmpty As Boolean = values Is Nothing OrElse values.Count() = 0

            Return If(isNullOrEmpty = False,
                      String.Format("{0}={1}",
                                    name,
                                    values.Aggregate(Function(previous, current) String.Format("{0},{1}", previous, current))
                                    ), Nothing)
		End Function

		Friend Shared Function GetFormattedUri(uriBase As String, ParamArray parameters As String()) As String
			Dim builder As New UriBuilder(uriBase)
			If parameters IsNot Nothing Then
                builder.Query = parameters.Aggregate(Function(previous, current) If(String.IsNullOrEmpty(previous),
                                                                                    current,
                                                                                    If(String.IsNullOrEmpty(current),
                                                                                       previous,
                                                                                       String.Format("{0}&{1}", previous, current))))
			End If
			Return builder.ToString()
		End Function

		Friend Shared Function DeserializeResponse(Of T)(uri As String, Optional method As String = "GET") As T
			Dim result As T
			Dim request As WebRequest = HttpWebRequest.Create(uri)
			request.Method = method
			Using response As WebResponse = request.GetResponse()
				Using stream As Stream = response.GetResponseStream()
					Dim serializer As New DataContractJsonSerializer(GetType(T))
					result = DirectCast(serializer.ReadObject(stream), T)
					stream.Close()
				End Using
				response.Close()
			End Using

			Return result
		End Function
	End Class
End Namespace
